---
title: "Predicting Salmonella, Listeria, Clostridium, and Eschrichia Genus and Species using Ribosomal 16S and 16-23S Intergenic Spacer Gene with Random Forest and Convolutional Neural Network (CNN)" 
author: "PiggyBotanist"
date: "2023-10-18"
output: pdf_document
---

# Introduction:

In the realm of medical diagnosis of infectious disease, the paramount initial step in formulating effective treatment plans revolves around the precise identification of the causative pathogens. Central to this identification process is the utilization of genetic markers, and among these, the 16S ribosomal RNA gene stands as a cornerstone, akin to the role of cytochrome C oxidase in the animal kingdom. The 16S gene's significance in bacterial species identification is widely acknowledged, and its application extends far and wide in the field of microbiology. Some common tests and analysis that revolves around 16S rRNA includes microbiome analysis, antibiotic resistance, PCR diagnostics, and more (Johnson et al., 2019, Clarridge. 2004, Rampini 2011).

	In addition to the 16S rRNA, recent research suggests the intergenic spacer region between 16S and 23S have promising properties to also aid in species identification, and can help enhance species resolution when combined with 16S rRNA (Tokajian et al., 2016, Stoppani et al., 2023) . Such resolution can offer the potential to distinguish between bacterial species at a higher resolution, which is crucial for distinguishing between commensal and pathogenic strains. Within the context of this assignment, I hope to explore the accuracy and efficacy of these promising genes, individually and in combination, for classifying the bacterial species, encompassing both commensal and pathogenic variants to provide better and more accurate diagnostic results for clinical microbiology.

# Background:

In order for such identification program to be useful in diagnostic labs, it has to be able to identify most pathogenic species with high accuracy and at times high accuracy to the serotype. Nonetheless, due to the time limitation of this assignment, I will only be exploring 4 of the most common pathogenic bacteria identified by government of food safety (www.foodsafety.gov), which would include clostridium, salmonella, escherichia(E. coli), and listeria. 

These four genus of bacteria are also well suited for this assignment because we have genus that are closely related, but also genus that are more distict. For example, salmonella and escherichia came from the same family (Enterobacteriaceae), while clostridium and listeria are from different order. Hence, this would allow us to perform classification based on data that are similar and distict between each other. 

In addition to genus that I am interested in, I will also be exploring the potential of classifying to serotype level because both salmonella and clostridium pathogenicity is strongly assicated with what serotypes they were. Some are more deadly then others.

# Section 1: Data Acquisition, Exploration, Filtering, and Quality Control.

## Data Acquisition:

Just as a reference, here are all the libraries that will be used throughout the script:
```{r setup}
library(stringr)
library(rentrez)
library(Biostrings)
library(randomForest)
library(dplyr)
library(ggplot2)
library(keras)
library(msa)
```

Since there are multiple species that I am interested in, I have defined a list that contains atomic vector within each position. Because I am interested in 16S as well as 16S-23S intergenic spacer, I will be downloading the sequences from NCBI (www.ncbi.nlm.nih.gov). Originally, I used a general search term for example(Clostridium[ORGN] AND 16S rRNA[Gene]), but this only yield a few sequences. Hence I decided to use more specific terms up to species level and are able to get way more datasets. Below are custom functions that I made for this assignment: 

```{r}
# Function called to combine searches and return dataframe
process_search_terms <- function(search_terms) {
  # Initialize an empty dataframe to store the results
  outputDataFrame <- data.frame()
  
  # Loop through each search term
  for (term in search_terms) {
    print(term[1])
    fileName <- "temp"
    search_detail <- term[2]
    
    # Obtain sequence data
    search_results <- entrez_search(db = "nuccore", term = search_detail, retmax = 100)
    id_list <- search_results$ids
    
    # Check if the ID list is empty
    if (length(id_list) == 0) {
      # Handle the case when no hits are found
      cat("No hits found for the given search term.\n")
    } else {
      sequences <- entrez_fetch(db = "nuccore", id = id_list, rettype = "fasta")
    
      # Write the sequences to the file
      writeLines(sequences, fileName)
    
      # Read the .fasta and store it as a dataframe
      file <- readDNAStringSet(fileName)
      file <- data.frame(
        original_id = names(file),
        id = word(names(file), 1L),
        classifier1 = word(names(file), 2L),
        classifier2 = word(names(file), 2L, 3L),
        nucleotide_length = nchar(paste(file)),
        nucleotides = paste(file)
      )
    
      # Combine the current data with the existing dataframe
      outputDataFrame <- rbind(outputDataFrame, file)
    }
  }
  # Remove variables created from the loop
  rm(fileName, search_detail, file)
  
  # Return the combined dataframe
  return(outputDataFrame)
}

# Function that iterates from 1-1000 decision trees and return a dataframe for accuracy and error rate at given parameter (version for classifier 1).
randomForest_ntreeIteration_1 <- function(dfTraining, dfValidation, nIterations = 1000) {
  result <- data.frame()
  
  for (i in 1:nIterations) {
    # Iterate from 1 to 1000, generate genus_classifier
    genus_classifier <- randomForest::randomForest(
      x = dfTraining,
      y = as.factor(dfTraining$classifier1),
      ntree = i,
      importance = TRUE
    )
    
    # find out of bag error rate
    oob_error_rate <- genus_classifier$err.rate[nrow(genus_classifier$err.rate), 1]
    
    # make predictions to the dataset
    dfValidation$prediction <- predict(genus_classifier, dfValidation)
    
    # obtain accuracy based on result
    accuracy <- sum(dfValidation$classifier1 == dfValidation$prediction) / nrow(dfValidation)
    
    # parse as a dataframe row, then bind it to our returning result
    new_row <- data.frame(
      ntree = i,
      error_rate = oob_error_rate,
      accuracy = accuracy
    )
    result <- bind_rows(result, new_row)
  }
  
  return(result)
}

# Function that iterates from 1-1000 decision trees and return a dataframe for accuracy and error rate at given parameter (version for classifier 2).
randomForest_ntreeIteration_2 <- function(dfTraining, dfValidation, nIterations = 1000) {
  result <- data.frame()
  
  for (i in 1:nIterations) {
    # Iterate from 1 to 1000, generate genus_classifier
    genus_classifier <- randomForest::randomForest(
      x = dfTraining,
      y = as.factor(dfTraining$classifier2),
      ntree = i,
      importance = TRUE
    )
    
    # find out of bag error rate
    oob_error_rate <- genus_classifier$err.rate[nrow(genus_classifier$err.rate), 1]
    
    # make predictions to the dataset
    dfValidation$prediction <- predict(genus_classifier, dfValidation)
    
    # obtain accuracy based on result
    accuracy <- sum(dfValidation$classifier1 == dfValidation$prediction) / nrow(dfValidation)
    
    # parse as a dataframe row, then bind it to our returning result
    new_row <- data.frame(
      ntree = i,
      error_rate = oob_error_rate,
      accuracy = accuracy
    )
    result <- bind_rows(result, new_row)
  }
  
  return(result)
}

# Function that iterates from 7-1000 (3 repeats) sample size and return a dataframe for accuracy and error rate at given parameter (version for classifier 1).
randomForest_sampleSizeIteration_1 <- function(dataset, nIterations = 21:3000, ntree = 250) {
  result <- data.frame()
  
  for (i in nIterations) {
    # define training dataset based on iteration
    dfTraining <- dataset[sample(nrow(dataset), size = ceiling(i/3)), ]
    dfValidation <- dataset %>% filter(!id %in% dfTraining$id)
    
    # define classifier
    genus_classifier <- randomForest::randomForest(
      x = dfTraining,
      y = as.factor(dfTraining$classifier1),
      ntree = ntree,
      importance = TRUE
    )
    # find out of bag error rate and accuarcy
    oob_error_rate <- genus_classifier$err.rate[nrow(genus_classifier$err.rate), 1]
    dfValidation$prediction <- predict(genus_classifier, dfValidation)
    accuracy <- sum(dfValidation$classifier1 == dfValidation$prediction) / nrow(dfValidation)
    
    # add data to result dataframe
    new_row <- data.frame(
      trainingSize = ceiling(i/3),
      error_rate = oob_error_rate,
      accuracy = accuracy
    )
    result <- bind_rows(result, new_row)
  }
  
  return(result)
}

# Function that iterates from 7-1000 (3 repeats) sample size and return a dataframe for accuracy and error rate at given parameter (version for classifier 2).
randomForest_sampleSizeIteration_2 <- function(dataset, nIterations = 21:3000, ntree = 250) {
  result <- data.frame()
  
  for (i in nIterations) {
    # define training dataset based on iteration
    dfTraining <- dataset[sample(nrow(dataset), size = ceiling(i/3)), ]
    dfValidation <- dataset %>% filter(!id %in% dfTraining$id)
    
    # define classifier
    genus_classifier <- randomForest::randomForest(
      x = dfTraining,
      y = as.factor(dfTraining$classifier2),
      ntree = ntree,
      importance = TRUE
    )
    # find out of bag error rate and accuarcy
    oob_error_rate <- genus_classifier$err.rate[nrow(genus_classifier$err.rate), 1]
    dfValidation$prediction <- predict(genus_classifier, dfValidation)
    accuracy <- sum(dfValidation$classifier1 == dfValidation$prediction) / nrow(dfValidation)
    
    # add data to result dataframe
    new_row <- data.frame(
      trainingSize = ceiling(i/3),
      error_rate = oob_error_rate,
      accuracy = accuracy
    )
    result <- bind_rows(result, new_row)
  }
  
  return(result)
}

```

```{r include=FALSE}
# Define NCBI search terms (for clostridium, salmonella, escherichia, and listeria with their 16S and 23S genes)

clostridium16S_terms <- list(
  c('carnis', "Clostridium carnis[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('fallax', "Clostridium fallax[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('haemolyticum', "Clostridium haemolyticum[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('histolyticum', "Clostridium histolyticum[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('bifermentans', "Clostridium bifermentans[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('ramosum', "Clostridium ramosum[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('sporogenes', "Clostridium sporogenes[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('tertium', "Clostridium tertium[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('subterminale', "Clostridium subterminale[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('butyricum', "Clostridium butyricum[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('akagii', "Clostridium akagii[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('sardiniense', "Clostridium sardiniense[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('botulinum', "Clostridium botulinum[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('difficile', "Clostridium difficile[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]")
)

escherichia16S_terms <- list(
  c('coli', "Escherichia coli[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('albertii', "Escherichia albertii[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('fergusonii', "Escherichia fergusonii[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]")
)

listeria16S_terms <- list(
  c('monocytogenes', "Listeria monocytogenes[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('grayi', "Listeria grayi[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('innocua', "Listeria innocua[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('ivanovii', "Listeria ivanovii[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('seeligeri', "Listeria seeligeri[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('welshimeri', "Listeria welshimeri[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]")
)

salmonella16S_terms <- list(
  c('typhimurium', "Salmonella typhimurium[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('enterica', "Salmonella enterica[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('enteritidis', "Salmonella enteritidis[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('newport', "Salmonella newport[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('paratyphi', "Salmonella paratyphi[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('virginia', "Salmonella virginia[ORGN] AND 16S NOT whole genome[All Fields] 200:2000[SLEN]")
)

clostridium23S_terms <- list(
  c('carnis', "Clostridium carnis[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('fallax', "Clostridium fallax[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('haemolyticum', "Clostridium haemolyticum[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('histolyticum', "Clostridium histolyticum[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('bifermentans', "Clostridium bifermentans[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('ramosum', "Clostridium ramosum[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('sporogenes', "Clostridium sporogenes[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('tertium', "Clostridium tertium[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('subterminale', "Clostridium subterminale[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('butyricum', "Clostridium butyricum[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('sardiniense', "Clostridium sardiniense[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('botulinum', "Clostridium botulinum[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('difficile', "Clostridium difficile[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]")
)

escherichia23S_terms <- list(
  c('coli', "Escherichia coli[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('albertii', "Escherichia albertii[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('fergusonii', "Escherichia fergusonii[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]")
)

listeria23S_terms <- list(
  c('monocytogenes', "Listeria monocytogenes[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('grayi', "Listeria grayi[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('innocua', "Listeria innocua[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('ivanovii', "Listeria ivanovii[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('seeligeri', "Listeria seeligeri[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('welshimeri', "Listeria welshimeri[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]")
)

salmonella23S_terms <- list(
  c('typhimurium', "Salmonella typhimurium[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('enterica', "Salmonella enterica[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('enteritidis', "Salmonella enteritidis[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('newport', "Salmonella newport[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('paratyphi', "Salmonella paratyphi[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]"),
  c('virginia', "Salmonella virginia[ORGN] AND (16S-23S OR 16S/23S) NOT whole genome[All Fields] 200:2000[SLEN]")
)

```

Two classifier was build into the function where classifier 1 will be the genus, and classifer 2 will be the genes-species. This will be used in the analysis for machine learning.

Using the function a total of 16 species of Clostridium, 3 species of escherichia, 6 species of listeria, and 6 species of salmonella were included. Within each search, the base pair length were restricted to between 200 and 2000 bps and non of them should include whole genome. The results were then retrieved using the follow script:

```{r}
clostridium16S <- process_search_terms(clostridium16S_terms)
escherichia16S <- process_search_terms(escherichia16S_terms)
listeria16S <- process_search_terms(listeria16S_terms)
salmonella16S <- process_search_terms(salmonella16S_terms)
clostridium23S <- process_search_terms(clostridium23S_terms)
escherichia23S <- process_search_terms(escherichia23S_terms)
listeria23S <- process_search_terms(listeria23S_terms)
salmonella23S <- process_search_terms(salmonella23S_terms)
```

## Exploration, filtering, and quality control:

After the all the sequences have been obtained, we should then first take a look at our data to make sure they are all 16S and 16S-23S intergenic spacer. 

Lets start with 16S datasets: I found out that there are a few rows where it contains tRNA or 16S-23S spacer, hence we will have to remove them first:

```{r}
# Remove rows containing tRNA for 16S
clostridium16S <- clostridium16S[grepl("16S", clostridium16S$original_id, ignore.case = TRUE), ]
escherichia16S <- escherichia16S[grepl("16S", escherichia16S$original_id, ignore.case = TRUE), ]
listeria16S <- listeria16S[grepl("16S", listeria16S$original_id, ignore.case = TRUE), ]
salmonella16S <- salmonella16S[grepl("16S", salmonella16S$original_id, ignore.case = TRUE), ]

clostridium16S <- clostridium16S[!grepl("tRNA|spacer", clostridium16S$original_id, ignore.case = TRUE), ]
escherichia16S <- escherichia16S[!grepl("tRNA|spacer", escherichia16S$original_id, ignore.case = TRUE), ]
listeria16S <- listeria16S[!grepl("tRNA|spacer", listeria16S$original_id, ignore.case = TRUE), ]
salmonella16S <- salmonella16S[!grepl("tRNA|spacer", salmonella16S$original_id, ignore.case = TRUE), ]
```

I also check the classifiers and some of them does not have consistent name, hence we only keep those that are consistent: 

```{r}
# Remove rows containing tRNA for 16S
clostridium16S <- clostridium16S[clostridium16S$classifier1 %in% c("Clostridium", "Clostridioides", "Erysipelatoclostridium", "Paraclostridium"), ]
escherichia16S <- escherichia16S[escherichia16S$classifier1 == "Escherichia", ]
listeria16S <- listeria16S[listeria16S$classifier1 == "Listeria", ]
salmonella16S <- salmonella16S[salmonella16S$classifier1 == "Salmonella", ]
```

Next we can start working on 23S datasets. It seems like there are also tRNA, however if we remove all that contains tRNA there will be little dataset, hence I decide to keep all as long as it contains the spacer.

```{r}
# Remove rows containing tRNA or 23S
clostridium23S <- clostridium23S[grepl("spacer", clostridium23S$original_id, ignore.case = TRUE), ]
escherichia23S <- escherichia23S[grepl("spacer", escherichia23S$original_id, ignore.case = TRUE), ]
listeria23S <- listeria23S[grepl("spacer", listeria23S$original_id, ignore.case = TRUE), ]
salmonella23S <- salmonella23S[grepl("spacer", salmonella23S$original_id, ignore.case = TRUE), ]
```

I also check the classifiers and some of them does not have consistent name, hence we only keep those that are consistent: 

```{r}
# Remove rows containing tRNA for 16S
clostridium16S <- clostridium16S[clostridium16S$classifier1 %in% c("Clostridium", "Clostridioides", "Paraclostridium"), ]
escherichia16S <- escherichia16S[escherichia16S$classifier1 == "Escherichia", ]
listeria16S <- listeria16S[listeria16S$classifier1 == "Listeria", ]
salmonella16S <- salmonella16S[salmonella16S$classifier1 == "Salmonella", ]
```

While I have not seen any "N"s in my sequences, I want to verify that I am correct. Hence I counted how many Ns appeared in my dataset using the follow script:

```{r}
clostridium16S$N_counts <- str_count(clostridium16S$nucleotides, "N")
escherichia16S$N_counts <- str_count(escherichia16S$nucleotides, "N")
listeria16S$N_counts <- str_count(listeria16S$nucleotides, "N")
salmonella16S$N_counts <- str_count(salmonella16S$nucleotides, "N")
clostridium23S$N_counts <- str_count(clostridium23S$nucleotides, "N")
escherichia23S$N_counts <- str_count(escherichia23S$nucleotides, "N")
listeria23S$N_counts <- str_count(listeria23S$nucleotides, "N")
salmonella23S$N_counts <- str_count(salmonella23S$nucleotides, "N")
```

Sure Enough, we found "N"s. Thus, I applied a filter where if N took 10% of the entire sequence, then it will be removed (Turns out none of them went above that):

```{r}
clostridium16S <- clostridium16S %>% filter(N_counts / nucleotide_length < 0.1)
escherichia16S <- escherichia16S %>% filter(N_counts / nucleotide_length < 0.1)
listeria16S <- listeria16S %>% filter(N_counts / nucleotide_length < 0.1)
salmonella16S <- salmonella16S %>% filter(N_counts / nucleotide_length < 0.1)
clostridium23S <- clostridium23S %>% filter(N_counts / nucleotide_length < 0.1)
escherichia23S <- escherichia23S %>% filter(N_counts / nucleotide_length < 0.1)
listeria23S <- listeria23S %>% filter(N_counts / nucleotide_length < 0.1)
salmonella23S <- salmonella23S %>% filter(N_counts / nucleotide_length < 0.1)
```

After filtering them we we got the following sequences:
- clostridium   16S:    707 -> 588 observations
- escherichia   16S:    300 -> 239 observations
- listeria      16S:    345 -> 312 observations
- salmonella    16S:    358 -> 311 observations
- clostridium   23S:    130 -> 126 observations
- escherichia   23S:    60 -> 58 observations
- listeria      23S:    40 -> 30 observations
- salmonella    23S:    44 -> 40 observations

Next I graphed the length of sequences onto a histogram to make sure there are no outliers (The codes were included but not graphed for the do to assignment limitations):

```{r eval=FALSE, include=FALSE}
# Show the distribution of my dataset to make sure there are not outliers
hist(nchar(clostridium16S$nucleotides),breaks = 100)      
hist(nchar(clostridium23S$nucleotides),breaks = 100)
hist(nchar(salmonella16S$nucleotides),breaks = 100)       
hist(nchar(salmonella23S$nucleotides),breaks = 100)    
hist(nchar(escherichia16S$nucleotides),breaks = 100)
hist(nchar(escherichia23S$nucleotides),breaks = 100)
hist(nchar(listeria16S$nucleotides),breaks = 100)
hist(nchar(listeria23S$nucleotides),breaks = 100)
```
Now that we have filtered the sequences, we should merged then align 16S and 23S to make it easier to analyze:
```{r}
merged_16S <- rbind(clostridium16S, escherichia16S, listeria16S, salmonella16S)
merged_23S <- rbind(clostridium23S, escherichia23S, listeria23S, salmonella23S)

merged_23S$aligned_seq <- as.character(msa(merged_23S$nucleotides, type = "DNA"))
# Alignment of 1450 observations would take a long time, hence only 23S was aligned.
#merged_16S$aligned_seq <- as.character(msa(merged_16S$nucleotides, type = "DNA"))


```

# Section 2: Using random forest to classify bacterial genus and species (Supervised machine learning):

## classification using 16S rRNA (Unaligned)

Since 16S is the more common gene used for identification, lets start with this dataset. To perform training, we first have to seperate the data into validation and training sets:

```{r}
set.seed(50)
dfValidation <- merged_16S %>% group_by(classifier1) %>% sample_n(5)

set.seed(20)
dfTraining <- merged_16S %>% filter(!id %in% dfValidation$id) %>% group_by(classifier1) %>% sample_n(22)
```

Now that we have the training dataset, let take a look at how number of decision tree might affect our error rate:

```{r}
decisionTreeResult_16S <- randomForest_ntreeIteration_1(dfTraining, dfValidation)

decisionTreeVSErrorRate_byUnique16S_genus <- ggplot(data = result, aes(x = ntree, y = error_rate)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, color = "blue") +  # Add a linear regression line
  labs(title = "Random Forest Error Rate (genus with 16S) vs. Number of Decision Trees", x = "Number of Trees", y = "Error Rate")

decisionTreeVSAccuracy_byUnique16S_genus <-ggplot(data = result, aes(x = ntree, y = accuracy)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, color = "blue") +  # Add a linear regression line
  labs(title = "Prediction Accuracy (genus with 16S) vs. Number of Decision Trees", x = "Number of Trees", y = "Accuracy")

```

From the first graph (decisionTreeVSErrorRate_byUnique16S), we can see that low numbers of decision tree causes high error rate, but as decision trees starts to increase beyond 250, the error seems to plateau at 0.0667. However, using a total of 110 sequences for training and 25 for validation  does not seem to be enough as we only have an accuracy of 60%. In the second graph (decisionTreeVSAccuracy_byUnique16S), we can see the number of trees does not seem to correlate with the accuracy what so ever. Hence, there is no need to go beyond ntree = 250. It is also counterintuitive where listeria and escherichia, as well as Paraclostridium and salmonella are being misidentified. Listeria and escherichia are different to the phylum level and salmonella and paraclostridium are from differe phylum which in theory they should be more distantly related. Where as those that are in the same family seems to have higher accuracy.

From Previous example I was test only 22 samples per genus, this time I would like to instead just randomly select 500 for training and 500 for validation regardless of the genus:

```{r}
# Define training and validation datasets
set.seed(50)
dfValidation <- merged_16S %>% sample_n(500)

set.seed(20)
dfTraining <- merged_16S %>% filter(!id %in% dfValidation$id) %>% sample_n(500)

# Run the function
decisionTreeResult_16S <- randomForest_ntreeIteration_1(dfTraining, dfValidation)

decisionTreeVSErrorRate_randomSelect16S_genus <- ggplot(data = decisionTreeResult_16S, aes(x = ntree, y = error_rate)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, color = "blue") +  # Add a linear regression line
  labs(title = "Random Forest Error Rate (genus with 16S) vs. Number of Decision Trees", x = "Number of Trees", y = "Error Rate")

decisionTreeVSAccuracy_randomSelect16S_genus <- ggplot(data = decisionTreeResult_16S, aes(x = ntree, y = accuracy)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, color = "blue") +  # Add a linear regression line
  labs(title = "Prediction Accuracy (genus with 16S) vs. Number of Decision Trees", x = "Number of Trees", y = "Accuracy")

```

Similar results was obtained were error rate and accuacy does not seem to be affected by the number of decision tree used from figure 3 and 4 (decisionTreeVSErrorRate_randomSelect16S, decisionTreeVSAccuracy_randomSelect16S). However, we can see that the accuracy is at 90% instead of the original 60%. Although some of the code have not been shown, I figure out that method 1 (where we select n samples based on their classifier) is much more effective than method 2 (just randomly select sequences) because at ntree = 250, with the same sample size for both traing and validation dataset, we found out method 1 yield an average of 60% accuracy while method only yield 52%. We are able to get a 90% instead of the 52% using method 2 here is purely because of the larger training dataset. Which can also be provide in the graph from the script below:

```{r}
sampleSizeResult_16S <- randomForest_sampleSizeIteration_1(merged_16S)

trainSizeVSAccuracy_randomSelect16S_genus <- ggplot(data = sampleSizeResult_16S, aes(x = trainingSize, y = accuracy)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, color = "blue") +  
  labs(title = "Prediction Accuracy (genus with 16S) vs. Training Size", x = "Training Dataset(n)", y = "Accuracy")
```

The sample thing was applied to classifer 2:

```{r}
# Define training and validation datasets
set.seed(50)
dfValidation <- merged_16S %>% sample_n(500)
set.seed(20)
dfTraining <- merged_16S %>% filter(!id %in% dfValidation$id) %>% sample_n(500)

# Run the function
decisionTreeResult_16S <- randomForest_ntreeIteration_2(dfTraining, dfValidation)

decisionTreeVSErrorRate_randomSelect16S_species <- ggplot(data = decisionTreeResult_16S, aes(x = ntree, y = error_rate)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, color = "blue") +  # Add a linear regression line
  labs(title = "Random Forest Error Rate (species with 16S) vs. Number of Decision Trees", x = "Number of Trees", y = "Error Rate")

decisionTreeVSAccuracy_randomSelect16S_species <- ggplot(data = decisionTreeResult_16S, aes(x = ntree, y = accuracy)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, color = "blue") +  # Add a linear regression line
  labs(title = "Prediction Accuracy (species with 16S) vs. Number of Decision Trees", x = "Number of Trees", y = "Accuracy")

sampleSizeResult_16S <- randomForest_sampleSizeIteration_2(merged_16S)

trainSizeVSAccuracy_randomSelect16S_species <- ggplot(data = sampleSizeResult_16S, aes(x = trainingSize, y = accuracy)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, color = "blue") +  
  labs(title = "Prediction Accuracy (species with 16S) vs. Training Size", x = "Training Dataset(n)", y = "Accuracy")

```

For the purpose of this assignment, I have only included four graphs that I find the most important, but all plots have been included in the script where one can always just plot the graph in R again to visualize it:

```{r}
plot_grid(decisionTreeVSAccuracy_byUnique16S_genus, trainSizeVSAccuracy_randomSelect16S_genus, decisionTreeVSAccuracy_byUnique16S_species, trainSizeVSAccuracy_randomSelect16S_species)
```

## classification using 23S rRNA (Unaligned)

```{r}
set.seed(50)
dfValidation <- merged_23S %>% sample_n(500)
set.seed(20)
dfTraining <- merged_23S %>% filter(!id %in% dfValidation$id) %>% sample_n(500)

# Run the function
decisionTreeResult_23S <- randomForest_ntreeIteration_1(dfTraining, dfValidation)

decisionTreeVSErrorRate_randomSelect23S_genus <- ggplot(data = decisionTreeResult_23S, aes(x = ntree, y = error_rate)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, color = "blue") +  # Add a linear regression line
  labs(title = "Random Forest Error Rate (species with 23S) vs. Number of Decision Trees", x = "Number of Trees", y = "Error Rate")

decisionTreeVSAccuracy_randomSelect23S_genus <- ggplot(data = decisionTreeResult_23S, aes(x = ntree, y = accuracy)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, color = "blue") +  # Add a linear regression line
  labs(title = "Prediction Accuracy (species with 23S) vs. Number of Decision Trees", x = "Number of Trees", y = "Accuracy")

sampleSizeResult_23S <- randomForest_sampleSizeIteration_1(merged_23S)

trainSizeVSAccuracy_randomSelect23S_genus <- ggplot(data = sampleSizeResult_23S, aes(x = trainingSize, y = accuracy)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, color = "blue") +  
  labs(title = "Prediction Accuracy (species with 23S) vs. Training Size", x = "Training Dataset(n)", y = "Accuracy")

# Run the function
decisionTreeResult_23S <- randomForest_ntreeIteration_2(dfTraining, dfValidation)

decisionTreeVSErrorRate_randomSelect23S_species <- ggplot(data = decisionTreeResult_23S, aes(x = ntree, y = error_rate)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, color = "blue") +  # Add a linear regression line
  labs(title = "Random Forest Error Rate (species with 23S) vs. Number of Decision Trees", x = "Number of Trees", y = "Error Rate")

decisionTreeVSAccuracy_randomSelect23S_species <- ggplot(data = decisionTreeResult_23S, aes(x = ntree, y = accuracy)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, color = "blue") +  # Add a linear regression line
  labs(title = "Prediction Accuracy (species with 23S) vs. Number of Decision Trees", x = "Number of Trees", y = "Accuracy")

sampleSizeResult_23S <- randomForest_sampleSizeIteration_2(merged_23S)

trainSizeVSAccuracy_randomSelect23S_species <- ggplot(data = sampleSizeResult_23S, aes(x = trainingSize, y = accuracy)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, color = "blue") +  
  labs(title = "Prediction Accuracy (species with 23S) vs. Training Size", x = "Training Dataset(n)", y = "Accuracy")
```

```{r}
plot_grid(decisionTreeVSAccuracy_byUnique23S_genus, trainSizeVSAccuracy_randomSelect23S_genus, decisionTreeVSAccuracy_byUnique23S_species, trainSizeVSAccuracy_randomSelect23S_species)
```

By applying the sample method from 16S to 23S we were able to get some predictions in. However, due to time constraints, I was only able to explore up to this point, but there are a lot more that can be explored to optimize prediction by using random forest, and I believe through fine tuning these parameters, we will be able to obtain a trained model that can have accuracy higher than 90%

# Section 3: Using neural network(keras) to classify bacterial genus and species (Supervised machine learning):

In addition to random forest, I would also like to perform convolutional nerual network (CNNs) to predict the species based on our sequence. However, there is a problem with our dataset that we need to solve before training it, which is the fact that we have variable sequence length. Hence I have decided to perform variable-length CNN first to see how it works:

```{r}

model <- keras_model_sequential()
```

After that, we can then train our neural network:
```{r}
set.seed(50)
dfValidation <- merged_16S %>% sample_n(500)
set.seed(20)
dfTraining <- merged_16S %>% filter(!id %in% dfValidation$id) %>% sample_n(500)

history <- model %>% fit(
  x = dfTraining$nucleotides,  # input data
  y = dfTraining$classifier2,  # target labels
  validation_data = list(dfValidation$nucleotides, dfValidation$classifier2),
  epochs = 10,  # number of training epochs
  batch_size = 50  # batch size
)


```


# Discussion:

Based on exploring the different parameters we found out that the number of decision trees does not really affect the accuracy, and anything above 250 reduced the OOB error rate to minimum which plateau at 0.67. I also compared using two method to train the random forest: 1) select specific size per unique classifier, 2) select random rows from the main datasets. It is worth noting that the group "clostridioides only contains 27 sequences, hence my method 1 is restricted to 22 for training and 5 for validation. Using method 1, approximately 60% accuracy was the highest I was able to get, and when using method 2 the accuracy was 52%. Hence, method 1 is definitely better than method 2. It is unfortunate that we don't have enough sample size to increase, and I believe if we have more samples per group we would yield a much higher accuracy

I also explored the effect of training set size vs accuracy and found out that 625 sequences was where the accuracy peaked (95%), and further increase in dataset size resulted in a drop in accuracy.


